File: packs/stack/python-fastapi/scaffold/app/infrastructure/repositories/memory_note_repository.py
from app.domain.entities.note import Note
from app.domain.interfaces.note_repository import NoteRepository


class MemoryNoteRepository(NoteRepository):
    """In memory implementation. Replace with a database repository when persistence arrives."""

    def __init__(self) -> None:
        self._notes: dict[str, Note] = {}

    async def add(self, note: Note) -> Note:
        self._notes[note.id] = note
        return note

    async def get(self, note_id: str) -> Note | None:
        return self._notes.get(note_id)

    async def list_all(self) -> list[Note]:
        return list(self._notes.values())

File: packs/stack/typescript-node/scaffold/src/infrastructure/memory-note-repository.ts
import { type Note, noteSchema } from "../domain/note.js";
import type { NoteRepository } from "../domain/note-repository.js";

/**
 * In memory implementation. Replace it with a database repository when persistence arrives, and
 * keep the noteSchema.parse call: raw storage data becomes an entity at this boundary or not at
 * all, so a malformed row fails loudly here instead of quietly inside a workflow.
 */
export class MemoryNoteRepository implements NoteRepository {
  readonly #notes = new Map<string, Note>();

  add(note: Note): Promise<Note> {
    const stored = noteSchema.parse(note);
    this.#notes.set(stored.id, stored);
    return Promise.resolve(stored);
  }

  get(id: string): Promise<Note | null> {
    const found = this.#notes.get(id);
    return Promise.resolve(found === undefined ? null : noteSchema.parse(found));
  }

  listAll(): Promise<Note[]> {
    return Promise.resolve([...this.#notes.values()].map((note) => noteSchema.parse(note)));
  }
}
